### PR TITLE
Give an error when using GameObject.gameObject #92

### LINKED ISSUE #92: Give an error when using GameObject.gameObject
# Problem statement
In relation to https://github.com/microsoft/Microsoft.Unity.Analyzers/issues/28, I just found out that `GameObject`s actually have a `gameObject` property which is undocumented but seems to just `return this`, so you can do `gameObject.gameObject.gameObject.gameObject...`.

And the same for `transform`.

# Proposed solution
Unless I'm missing something, it seems to do absolutely nothing so it should give an error any time you use it.

### PR DESCRIPTION
**Pull request by shreyalpandit**
Fixes #92

#### Short description of what this resolves:
GameObjects have a gameObject property which is undocumented but returns this, so you can do gameObject.gameObject.gameObject.gameObject...

#### Changes proposed in this pull request:
-New Analyzer to give a message for gameObject indirection

### CODE REVIEW

File: src/Microsoft.Unity.Analyzers/IndirectionMessage.cs [hunk 1]
```diff
@@ -0,0 +1,110 @@
+/*--------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See LICENSE in the project root for license information.
+ *-------------------------------------------------------------------------------------------*/
+
+//using System.Collections.Immutable;
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers/IndirectionMessage.cs:6]: We can remove those commented namespaces + those are not used:

```csharp
using System;
using System.Collections.Generic;
using System.Linq;
```

File: src/Microsoft.Unity.Analyzers/IndirectionMessage.cs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+		internal static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+			id: "UNT0019",
+			title: Strings.IndirectionMessageDiagnosticTitle,
+			messageFormat: Strings.IndirectionMessageDiagnosticMessageFormat,
+			category: DiagnosticCategory.Performance,
+			defaultSeverity: DiagnosticSeverity.Info,
+			isEnabledByDefault: true,
+			description: Strings.IndirectionMessageDiagnosticDescription);
+
+		internal const string UpdateId = "UNT0019";
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers/IndirectionMessage.cs:42]: unused

File: src/Microsoft.Unity.Analyzers/IndirectionMessage.cs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+			if (!(context.Node is MemberAccessExpressionSyntax))
+				return;
+
+			var access = (MemberAccessExpressionSyntax)context.Node;
+			var symbol = context.SemanticModel.GetSymbolInfo(access);
+
+			if (symbol.Symbol == null)
+				return;
+
+			if (!(symbol.Symbol is IPropertySymbol property))
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers/IndirectionMessage.cs:65]: property is unused after so no need to use the full pattern matching syntax:

```csharp
if (!(symbol.Symbol is IPropertySymbol))
				return;
```

should be enough?

File: src/Microsoft.Unity.Analyzers/IndirectionMessage.cs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+				context.Diagnostics);
+		}
+
+		private static async Task<Document> DeleteIndirectionAsync(Document document, MemberAccessExpressionSyntax access, CancellationToken cancellationToken)
+		{
+			var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+			var newExpression = access.Expression;
+			var newRoot = root.ReplaceNode(access,newExpression);
+
+			return document.WithSyntaxRoot(newRoot);
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers/IndirectionMessage.cs:106]: WithSyntaxRoot checks for a not null root.

I think your code is ok (we should never have something null from root.ReplaceNode) but static analysis is complaining.

So perhaps we can do something like that:
```csharp
if (newRoot == null)
	return document;
	
return document.WithSyntaxRoot(newRoot);
```

File: src/Microsoft.Unity.Analyzers.Tests/ContextMenuSuppressorTests.cs [hunk 5]
```diff
@@ -137,4 +137,4 @@ static void DoMenuItem()
 		}
 
 	}
-}
+}
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers.Tests/ContextMenuSuppressorTests.cs:140]: Unneeded change. Our editorconfig is enforcing a new line at the end of all text files if I'm not wrong

File: src/Microsoft.Unity.Analyzers/Resources/Strings.resx [hunk 6]
```diff
@@ -384,4 +384,16 @@
...
+  <data name="IndirectionMessageDiagnosticDescription" xml:space="preserve">
+    <value>Each access is called.</value>
+  </data>
+  <data name="IndirectionMessageDiagnosticMessageFormat" xml:space="preserve">
+    <value>Indirection is redundant</value>
+  </data>
+  <data name="IndirectionMessageDiagnosticTitle" xml:space="preserve">
+    <value>Indirection call</value>
+  </data>
+</root>
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers/Resources/Strings.resx:399]: Not super important but we should keep the newline after `</root>`

**Comment by sailro** [comment]
@shreyalpandit did you try outside the test project  ? On my side I have an issue with a real Unity project:

```
Warning	CS8032	An instance of analyzer Microsoft.Unity.Analyzers.IndirectionMessageAnalyzer cannot be created from C:\Dev\Microsoft.Unity.Analyzers-intern\src\Microsoft.Unity.Analyzers\bin\Debug\netstandard2.0\Microsoft.Unity.Analyzers.dll: Could not load type 'Microsoft.Unity.Analyzers.IndirectionMessageAnalyzer' from assembly 'Microsoft.Unity.Analyzers, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'..	Assembly-CSharp		1	Active	System.TypeLoadException: Could not load type 'Microsoft.Unity.Analyzers.IndirectionMessageAnalyzer' from assembly 'Microsoft.Unity.Analyzers, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'.
   at System.Reflection.RuntimeAssembly.GetType(RuntimeAssembly assembly, String name, Boolean throwOnError, Boolean ignoreCase, ObjectHandleOnStack type)
   at System.Reflection.RuntimeAssembly.GetType(String name, Boolean throwOnError, Boolean ignoreCase)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerFileReference.Extensions`1.GetAnalyzersForTypeNames(Assembly analyzerAssembly, IEnumerable`1 analyzerTypeNames, Boolean& reportedError)
```

Other analyzers are loaded:

image

File: src/Microsoft.Unity.Analyzers/IndirectionMessage.cs [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+			var access = (MemberAccessExpressionSyntax)context.Node;
+			var symbol = context.SemanticModel.GetSymbolInfo(access);
+
+			if (symbol.Symbol == null)
+				return;
+
+			if (!(symbol.Symbol is IPropertySymbol property))
+				return;
+
+			if (!context.SemanticModel.GetTypeInfo(access.Expression).Type.Extends(typeof(UnityEngine.GameObject)))
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers/IndirectionMessage.cs:68]: Perhaps we should introduce a variable for GetTypeInfo. And TypeInfo.Type can be null with malformed syntax.

Comment (sailro) [reply src/Microsoft.Unity.Analyzers/IndirectionMessage.cs:68]: That was an issue on my side. Closing (I had an old vsix with the analyzer dll)

**Comment by sailro** [comment]
LGTM, you just need to add documentation and address minor comments :)

See doc directory. We need to add a `UNT0019.md` file and reference it from the `index.md` 

Thank you @shreyalpandit !

**Review by shreyalpandit — COMMENTED** [review]
Fixed comments and added documentation

File: doc/UNT0019.md [hunk 8]
```diff
@@ -0,0 +1,39 @@
+# UNT0019 Unnecessary indirection call for GameObject.gameObject and Transform.transform
+
+Unity `GameObject` has a property `gameObject` that returns `this` and same for
```
Comment (sailro) [inline doc/UNT0019.md:3]: Indeed `GameObject` has a property `gameObject`returning `this`. But for Transform it is a bit different.
Transform inherits from `Component` and a `Component` exposes `gameObject` and `transform` properties. Of course Transform.transform will return `this`. In the end we are looking for the same unnecessary indirection. Do you plan to support this case in another PR ?

**Comment by sailro** [comment]
Thank you @shreyalpandit, we can squash/merge this PR.

**Review by sailro — APPROVED** [review]

File: doc/UNT0019.md (hunk 8 above)
Comment (shreyalpandit) [reply doc/UNT0019.md:3]: Yes, I just removed the reference to Transform and will do this in another PR.

Comment (sailro) [reply doc/UNT0019.md:3]: Perfect. Thanks.

**Review by shreyalpandit — COMMENTED** [review]
removed transform from documentation

File: src/Microsoft.Unity.Analyzers/IndirectionMessage.cs [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+			if (!(symbol.Symbol is IPropertySymbol))
+				return;
+				
+			if (typeInfo.Type == null)
+				return;
+				
+			if (!typeInfo.Type.Extends(typeof(UnityEngine.GameObject)))
+				return;
+
+			if (access.Name.ToFullString() != "gameObject")
```
Comment (jbevain) [inline src/Microsoft.Unity.Analyzers/IndirectionMessage.cs:65]: Please move this between the declaration of `var access` and accessing the semantic model.

The reason for that is that accessing the SemanticModel is time confusing, whereas comparing the name of the member being access is really fast. This way we're only going to do the costly analysis for members named `gameObject`.

**Comment by jbevain** [comment]
Hi @shreyalpandit. Thank you for the PR, great work. Please just address my comment and this is good to go. Thanks!

**Review by shreyalpandit — COMMENTED** [review]
Moved check for name before accessing semantic model.

**Pull request merged**